man/man1/pcp-shping.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2015,2019 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PCP-SHPING 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pcp-shping\f1 \- report on shell service availability and response
.SH SYNOPSIS
\f3pcp\f1
[\f2pcp\ options\f1]
\f3shping\f1
[\f3\-cl?\f1]
\f2tag\f1 ...
.SH DESCRIPTION
.B pcp-shping
samples and reports on the shell-ping service metrics exported
by the
.BR pmdashping (1)
agent.
.PP
The default report from
.B pcp-shping
shows two columns for each service
.IR tag ,
the first showing service status (zero indicating success) and the
second service response time, for the last
.B pmdashping
command refresh cycle.
.PP
When invoked via the
.BR pcp (1)
command, the
.BR \-h /\c
.BR \-\-host ,
.BR \-a /\c
.BR \-\-archive ,
.BR \-O /\c
.BR \-\-origin ,
.BR \-s /\c
.BR \-\-samples ,
.BR \-t /\c
.BR \-\-interval ,
.BR \-Z /\c
.BR \-\-timezone
and several other
.B pcp
.I options
become indirectly available, see
.BR PCPIntro (1)
for their descriptions.
.SH OPTIONS
The available command line options are:
.TP 5
\fB\-c\fP, \fB\-\-status\fP
Display a summary of the current shell-ping collector, including
information about service refresh interval, the current command
timeout setting, count of refreshes, and the full command
line associated with each
.IR tag .
.TP
\fB\-l\fP, \fB\-\-tags\fP
Displays the valid
.IR tag s.
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pcp (1)
and
.BR pmdashping (1).

.\" control lines for scripts/man-spell
.\" +ok+ shping [from pcp-shping]


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/pcp/shping/pcp-shping.sh
#! /bin/sh
# 
# Copyright (c) 2015 Red Hat.
# 
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# Displays stats from the shell ping Performance Co-Pilot domain agent.
#

. $PCP_DIR/etc/pcp.env

sts=2
progname=`basename $0`
tmp=`mktemp -d "$PCP_TMPFILE_DIR/pcp-shping.XXXXXXXXX"` || exit 1
trap "rm -rf $tmp; exit \$sts" 0 1 2 3 15

_usage()
{
    [ ! -z "$@" ] && echo $@ 1>&2
    pmgetopt --progname=$progname --usage --config=$tmp/usage
    exit 1
}

cat > $tmp/usage << EOF
# Usage: [options] tag...

Options:
   -l,--tags	report the list of shell-ping command tags
   -c,--status	report shell-ping service check control variables
   --help
# end
EOF

cflag=false
lflag=false
ARGS=`pmgetopt --progname=$progname --config=$tmp/usage -- "$@"`
[ $? != 0 ] && exit 1

eval set -- "$ARGS"
while [ $# -gt 0 ]
do
    case "$1" in
      -c)
	cflag=true
	;;
      -l)
	lflag=true
	;;
      -\?)
	_usage ""
	;;
      --)	# end of options, start of arguments
	shift
	while [ $# -gt 0 ]
	do
	    echo \"$1\" >> $tmp/tags
	    shift
	done
	break
	;;
    esac
    shift	# finished with this option now, next!
done

pmprobe -I shping.cmd | grep -v 'Note: timezone' > $tmp/cmd 2> $tmp/err
if grep "^pmprobe:" $tmp/err > /dev/null 2>&1
then
    $PCP_ECHO_PROG $PCP_ECHO_N "$progname: ""$PCP_ECHO_C"
    sed < $tmp/err -e 's/^pmprobe: //g'
    sts=1
    exit
fi

set `cat $tmp/cmd`
if [ "$2" -eq "-12357" ]	# Unknown metric name
then
    $PCP_ECHO_PROG "$progname: pmdashping(1) is not installed"
    sts=1
    exit
elif [ "$2" -lt 0 ]		# Some other PMAPI error
then
    shift && shift
    $PCP_ECHO_PROG "$progname: $*"
    sts=1
    exit
elif [ "$2" -eq 0 ]		# No instances?
then
    $PCP_ECHO_PROG "$progname: no shell ping command instances"
    sts=1
    exit
fi
shift && shift		# skip over name and error/count

if $lflag
then
    $PCP_ECHO_PROG "$progname: $*"
    sts=0
    exit
elif $cflag
then
    $PCP_ECHO_PROG $PCP_ECHO_N "Duty cycles:       ""$PCP_ECHO_C"
    pmprobe -v shping.control.cycles | $PCP_AWK_PROG '{ print $3 }'
    $PCP_ECHO_PROG $PCP_ECHO_N "Refresh interval:  ""$PCP_ECHO_C"
    pmprobe -v shping.control.cycletime | $PCP_AWK_PROG '{ print $3, "secs" }'
    $PCP_ECHO_PROG $PCP_ECHO_N "Check timeout:     ""$PCP_ECHO_C"
    pmprobe -v shping.control.timeout | $PCP_AWK_PROG '{ print $3, "secs" }'
    $PCP_ECHO_PROG
    $PCP_ECHO_PROG "Tags and command lines:"
    pminfo -f shping.cmd | \
	sed -e '/^$/d' -e '/^shping/d' \
	    -e 's/"] value/]/' -e 's/^ *inst//' \
	    -e 's/[0-9][0-9]* or "//g'
    sts=0
    exit
fi

# positional args now hold shping metric instance names
while [ $# -gt 0 ]
do
    if [ -s $tmp/tags ]
    then
	if ! grep -q "^$1$" $tmp/tags
	then
	    shift
	    continue
	fi
    fi
    # build up a pmie configuration file
    cmd=`echo $1 | tr -d '"'`
    echo "'sts_$cmd' = shping.status #'$cmd';" >> $tmp/pmie
    echo "'time_$cmd' = shping.time.real #'$cmd';" >> $tmp/pmie
    shift
done

if [ ! -s $tmp/pmie ]
then
    $PCP_ECHO_PROG "$progname: no matching command tags found"
    sts=1
    exit
fi

cat $tmp/pmie | pmie -v -e -q | pmie2col -p 3 -w 10
sts=0
exit
